src/libultra_code/contpfs.c

```diff
@@ -0,0 +1,318 @@
+            return ret;
+        }
+        temp[0] = j | 0x80;
+        for (i = 1; i < BLOCKSIZE; i++) {
+            temp[i] = ~temp[i];
+        }
+
+        if ((ret = __osContRamWrite(pfs->queue, pfs->channel, 0, temp, 0)) != 0) {
+            return ret;
+        }
+        if ((ret = __osContRamRead(pfs->queue, pfs->channel, 0, comp)) != 0) {
+            return (ret);
+        }
+        for (i = 0; i < BLOCKSIZE; i++) {
+            if (comp[i] != temp[i]) {
+                break;
+            }
+        }
+        if (i != BLOCKSIZE) {
+            break;
+        }
+
+        if (j > 0) {
+            if ((ret = __osPfsSelectBank(pfs, 0)) != 0) {
+                return ret;
+            }
+            if ((ret = __osContRamRead(pfs->queue, pfs->channel, 0, temp)) != 0) {
+                return ret;
+            }
+            if (temp[0] != 0x80) {
+                break;
+            }
+        }
+
+        j++;
+    } while (j < PFS_MAX_BANKS);
+
+    if ((pfs->activebank != 0) && (ret = __osPfsSelectBank(pfs, 0)) != 0) {
+        return ret;
+    }
+
+    mask = (j > 0) ? 1 : 0;
+    newid->deviceid = (badid->deviceid & 0xFFFE) | mask;
+    newid->banks = j;
+    newid->version = badid->version;
+    __osIdCheckSum(newid, &newid->checksum, &newid->invertedChecksum);
+
+    index[0] = PFS_ID_0AREA;
+    index[1] = PFS_ID_1AREA;
+    index[2] = PFS_ID_2AREA;
+    index[3] = PFS_ID_3AREA;
+    for (i = 0; i < 4; i++) {
+        if ((ret = __osContRamWrite(pfs->queue, pfs->channel, index[i], newid, PFS_FORCE)) != 0) {
+            return ret;
+        }
+    }
+    if ((ret = __osContRamRead(pfs->queue, pfs->channel, PFS_ID_0AREA, temp)) != 0) {
+        return ret;
+    }
+    for (i = 0; i < BLOCKSIZE; i++) {
+        if (temp[i] != *(u8*)((s32)newid + i)) { // Cast to s32 required to match here
+            return PFS_ERR_DEVICE;
+        }
+    }
+    return 0;
+}
+
+s32 __osCheckPackId(OSPfs* pfs, __OSPackId* temp) {
+
+    u16 index[4];
+    s32 ret = 0;
+    u16 sum;
+    u16 idSum;
+    s32 i;
+    s32 j;
+
+    if ((pfs->activebank != 0) && (ret = __osPfsSelectBank(pfs, 0)) != 0) {
+        return ret;
+    }
+
+    index[0] = PFS_ID_0AREA;
+    index[1] = PFS_ID_1AREA;
+    index[2] = PFS_ID_2AREA;
+    index[3] = PFS_ID_3AREA;
+    for (i = 1; i < 4; i++) {
+        if ((ret = __osContRamRead(pfs->queue, pfs->channel, index[i], temp)) != 0) {
+            return ret;
+        }
+        __osIdCheckSum(temp, &sum, &idSum);
+        if ((temp->checksum == sum) && (temp->invertedChecksum == idSum)) {
+            break;
+        }
+    }
+    if (i == 4) {
+        return PFS_ERR_ID_FATAL;
+    }
+
+    for (j = 0; j < 4; j++) {
+        if (j != i) {
+            if ((ret = __osContRamWrite(pfs->queue, pfs->channel, index[j], temp, PFS_FORCE)) != 0) {
+                return ret;
+            }
+        }
+    }
+    return 0;
+}
+
+#ifdef NON_MATCHING
+// functionally equivalent, just regalloc with temps and v1/v0
```

Comment:
functional equivalence is implied when using NON_MATCHING instead of NON_EQUIVALENT, so you can just state the problem